### ISSUE TITLE
Specifying the order of virtual tracks

### ISSUE THREAD

**Issue by dtatut** [imfug/delivery-schema#14]
Some businesses require/mandate that the virtual tracks must be present in the CPL in a specific order. Currently this is not really possible or obvious (initially the order of presence in the delivery schema was suggested, but that was too weak). Therefore I suggest the following:

- for the VirtualTrackList element, add an attribute called "sequence"
- that attribute shall have two possible values: unordered and ordered
- the default value for the sequence attribute shall be "unordered"

**Comment by fschleich** [comment]
I would prefer to learn more about the need for this and the use-case first, before discussing solutions to a problem that I don't understand. Could some of those businesses elaborate?

**Comment by palemieux** [comment]
> Could some of those businesses elaborate?

+1 

I am having a hard time thinking of a reason to require a specific Virtual Track ordering since ordering is not used/enforced by the specifications.

**Comment by dtatut** [comment]
@fschleich @palemieux : This problem is not related to the way IMF CPLs specify tracks, but unfortunately related to the way some "IMF-ready" tools (available in the market) expect tracks to be arranged. Moreover, some studios seem to have expectations on the track order as well.

For the upcoming call, Dave Kneeland/Fox is supposed to clarifiy the position of his team on this matter.

In general this problem is only related to the audio tracks. The real question is how to clearly identify the presence (or absence) of a certain audio virtual track, especially when the soundfield group and channel layout identical across several tracks. Obviously a good way would be to make a better use of the MCA metadata. Unfortunately, so far, we only had very limited feedback.

Any thoughts, feedback and alternatives are obviously welcome.

**Comment by palemieux** [comment]
> related to the way some "IMF-ready" tools (available in the market) expect tracks to be arranged.

Thanks for the clarification. It would be good to understand if this is related to a deficiency in the specifications, or merely a bug in a tool.

**Comment by dtatut** [comment]
@palemieux : most of the time it's a limitation of the tool. But not only. In general the order of the audio tracks (often confused with audio channels) follows what I like to call a "natural order". Delivery specifications are in most cases, highly inspired by their defunct (crossing fingers!) counterparts for tape-based workflows.

My problem here is how can we help with the transition process. My proposal above is meaningless for modern workflows, especially for proper IMF implementations. Nevertheless, I cannot ignore the reality and this transition that will certainly take longer than expected.

While I agree with @fschleich and yourself, the proposal above is only to be used by those who need it. If a studio does not want to rely on such legacy, then the studio can simply omit the attribute, as explained. So the attribute is not impacting everyone, but only those who need a specific track layout.

**Comment by fschleich** [comment]
> My problem here is how can we help with the transition process. My proposal above is meaningless for modern workflows, especially for proper IMF implementations. Nevertheless, I cannot ignore the reality and this transition that will certainly take longer than expected.

My concern is that giving this some "legitimacy" by supporting it in the schema might lead users to specifying track orders simply because they don't know better. I don't want to miss the opportunity to discuss this with the folks that rely on an ordered sequence, to make sure this requirement is based on a real need and not misunderstandings.

**Comment by palemieux** [comment]
Can a test vector be designed to trigger this behavior? It could included in an upcoming plugfest.

In the meantime I would resist enshrining a behavior that fundamentally conflicts with object-based / component-based media.

**Comment by MarcAntoine-Arnaud** [comment]
Do you think it can start a discussion to generate some exhaustive CPL cases to provide a list of required support (IMF tools can build their unit tests with them).
It's already done for lot of W3C specifications (like IMSC, RDF, etc.)

**Comment by dtatut** [comment]
@fschleich : I think we need to create that oppotunity and we have a platform for doing this: the IMFUG. Let's bring up this topic and ask the members at the next UG session. Maybe not to get a final decision, but at least to ask them to collect feedback for the session after.

@palemieux : yes definitely. Also since this is essentially related to audio tracks, I would suggest to combine this with (apparently) a necessary test vector on MCA.

@MarcAntoine-Arnaud : like I said above, IMFUG seems the right platform for me.

What do you think?

**Comment by dkneeland** [comment]
> For the upcoming call, Dave Kneeland/Fox is supposed to clarifiy the position of his team on this matter.

I don’t think that type of check should be within the Delivery Schema, due to the fact that the IMF standard does not mandate what the order of tracks should be. If we decide as a group that we do care about the track order, then it should be added to the standard first and then inherited by the Delivery Schema.

With that said, I do agree with @dtatut that the track order is an important thing due to legacy IMF implementations, and that this should be a conversation we should have at the next IMFUG.

**Comment by jhursty** [comment]
+1

-jh

**Comment by ghost** [comment]
+1

Much as it pains me, for rolling IMF into broadcast workflows, we have to accept that some legacy interfaces are going to persists for a while. For this reason some of our members will ask for a certain order of virtual tracks. If you all say "no" I will be happy in a way. But I have to ask on behalf of our members ;-)

**Comment by palemieux** [comment]
> we have to accept that some legacy interfaces are going to persists for a while.

I am not sure how legacy interfaces relate to ordering of Virtual Tracks: the IMF implementations can reorder prior to output to legacy interfaces.

The issue suggested at https://github.com/imfug/delivery-schema/issues/14#issuecomment-475176506 is the existence of legacy _IMF implementations_ :(

**Comment by dtatut** [comment]
Based on the discussions here and those had during the tech call today, I'm removing this issue from the NAB2019 milestone.

The decision was made to take this subject before the IMFUG members to discuss the need for ordered tracks.

Currently, the version for the NAB2019 milestone will only rely on the MCA framework.

**Comment by ghost** [comment]
Hi Pierre,

> The issue suggested at [#14 (comment)](https://github.com/imfug/delivery-schema/issues/14#issuecomment-475176506) is the existence of legacy _IMF implementations_ :(

Yes and no, in my view :-)
If we want our members to adopt IMF we have to bring some bridge building material. There are many legacy workflows which will not be easily uprooted (whether that is based on long-running supply contracts with third parties for localisation, TX, etc. or simply on internal workflows which we are chipping away at but which will not change in one fell swoop). Wholesale drop-in of component-based working across the supply chain is not going to happen in our use cases.

But IMF is not fully automatable at this moment - you are aware of our efforts to create that nirvana with the help of OPL, but this is not going to finish very soon, we are just scratching the surface.
So, being able to specify a sound-field order will allow some implementers to drop in IMF packages in place of their current flat files for transcoding and onward processing.

Yes, MCA labels should be all that is needed, and yes, I understand that the provision in the DS for track order may create a legacy issue for us... but I am simply highlighting that some of our members want to see track order specifications.

**Comment by jhursty** [comment]
Ulrich,

I feel compelled to point out that the audio routing macro is alive and well [0], and provides a working, normative solution to the channel ordering problem.

I get that many are waiting for some as-yet unaddressed image processing, but this deficiency in the picture domain should not occlude the ready-to-run solution to this very particular audio issue.

Perhaps the audio routing macro should become the focus of some near-term plug-festing? IMO, this would take less work and will certainly produce a better result that retrofitting track order to the CPL.


Best,

-jh

[0] ST 2067-103:2014 https://doi.org/10.5594/SMPTE.ST2067-103.2014

**Comment by dtatut** [comment]
@Ulrich-DPP : I know it's a bit of a short notice, but would it be possible to collect some rapid feedback from your members for the upcoming discussion at the next IMFUG?

@jhursty : that implies OPL processing. While you are correct on this, the adoption of IMF has been done without OPLs for the past years, ... It will take another couple of years before proper OPL processing solves this problem here.

**Comment by mrmxf** [comment]
I'm attracted to the OPL solution. One of the problems of a deterministic
virtual track order is that it can't be right for everyone. The track order
for a given CPLId  is immutable. Those two facts sit uncomfortably together
in my brain. An immutable CPL order and a bendable, customer specific order
(MCA tags to ordering) is actually quite elegant. This could be a CPL to
CPL operation where the new CPL is essentially a disposable working
artefact of the original Immutable CPL. Not only does this give you a
workflow solution that you can use, it's easy to QC, easy to code and it's
easy to delete when it becomes obsolete.

Makes sense to me (which probably means it's a bit  ... out there)

*Bruce*

------------------------------
T:+44 777 6464 888. Mr MXF Ltd. <http://mrmxf.com/> Media consultancy,
standards, education, project incantation & management.
Try IMFTool <https://github.com/IMFTool>, Wolfgang will thank you.
------------------------------
<https://mrmxf.com>

**Comment by ghost** [comment]
@dtatut Of course - we will present this on Monday in our IMF Stakeholders' Group.

@mrmxf I like the synthesis. After all, CPL are essentially 'free' ;)
Let's see where this goes...

**Comment by fschleich** [comment]
@mrmxf @Ulrich-DPP While this is not the topic of this thread, I just want to point out that channel ordering in actual Sound Track Files can be all over the place, thanks to MCA labeling. A CPL based Virtual Track reshuffle would only cut it if the targeted "legacy implementation" was reliably MCA aware (which I would question).

**Comment by mrmxf** [comment]
Agreed - that's why I think it's an OPL issue. You'd have to rewrap the
track file into something that was in the concrete order required. (in
addition to doing the CPL). The full set of requirements is much more than
metadata flipping :-)

------------------------------
T:+44 777 6464 888. Mr MXF Ltd. <http://mrmxf.com/> Media consultancy,
standards, education, project incantation & management.
Try IMFTool <https://github.com/IMFTool>, Wolfgang will thank you.
------------------------------
<https://mrmxf.com>
